inst/include/frogger.hpp

```diff
@@ -32,32 +34,27 @@ void initialise_model_state(const Parameters<real_type> &pars,
 
 }
 
-template<typename real_type>
-typename StateSaver<real_type>::OutputState run_model(int time_steps, std::vector<int> save_steps,
-                                                      const Parameters<real_type> &pars) {
-  State<real_type> state(pars.age_groups_pop, pars.num_genders,
-                         pars.disease_stages, pars.age_groups_hiv,
-                         pars.treatment_stages);
+template<typename real_type, HivAgeStratification S>
+typename StateSaver<real_type, S>::OutputState run_model(int time_steps,
+                                                         std::vector<int> save_steps,
+                                                         const Parameters<real_type> &pars) {
+  State<real_type, S> state;
 
-  internal::initialise_model_state(pars, state);
+  internal::initialise_model_state<real_type, S>(pars, state);
   auto state_next = state;
-  internal::IntermediateData<real_type> intermediate(pars.age_groups_pop, pars.age_groups_hiv, pars.num_genders,
-                                                     pars.disease_stages, pars.treatment_stages,
-                                                     pars.age_groups_hiv_15plus);
+  internal::IntermediateData<real_type, S> intermediate(pars.age_groups_hiv_15plus);
   intermediate.reset();
 
-  StateSaver<real_type> state_output(time_steps, save_steps, pars.age_groups_pop, pars.num_genders,
-                                     pars.disease_stages, pars.age_groups_hiv,
-                                     pars.treatment_stages);
+  StateSaver<real_type, S> state_output(time_steps, save_steps);
   // Save initial state
   state_output.save_state(state, 0);
 
   // Each time step is mid-point of the year
   for (int step = 1; step <= time_steps; ++step) {
     state_next.reset();
-    run_general_pop_demographic_projection(step, pars, state, state_next, intermediate);
-    run_hiv_pop_demographic_projection(step, pars, state, state_next, intermediate);
-    run_hiv_model_simulation(step, pars, state, state_next, intermediate);
+    run_general_pop_demographic_projection<real_type, S>(step, pars, state, state_next, intermediate);
```

Comment:
You probably don't want to hear this, but I think that if you reorder the template parameters you may be able to drop the `real_type` here and just template explicitly on `S`. I forget if this works with mixed type/literal templates though

Reply:
Sounds good, I've tidied that up